File: A10 Recursion, DFS, Backtracking, BFS/LC700_Search In Binary Tree/SearchBST.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution(object):
  def search(self, root, key):
    """
    input: TreeNode root, int key
    return: TreeNode
    """
    # write your solution here


#####Recursive approoach####
def recursive(self, root, val):
        def rec(root):
            if root:
                if root.val == val: return root
                elif root.val < val: return rec(root.right)
                return rec(root.left)
        
        return rec(root)


######iterative Approach#####
def iterative(self, root, val):
        temp = root
        while temp:
            if temp.val == val: return temp
            elif temp.val < val: temp = temp.right
            else: temp = temp.left
        return None

File: A10 Recursion, DFS, Backtracking, BFS/LC700_Search In Binary Tree/SearchBST.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        TreeNode curr = root; 
        while (curr != null){
            if(curr.val == val){
                return curr; 
            }
            else if(curr.val > val){
                curr = curr.left; 
            }
            else
                curr = curr.right;    
        }
        return curr;
        
    }
}

##########Recursion#######
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode searchBST(TreeNode root, int val) {
        if(root == null)
            return root;
        if (root.val > val)
            root = searchBST(root.left, val);
        else if (root.val < val)
            root = searchBST(root.right, val);
        
        return root;   
        
    }
}


########################
    public TreeNode searchBST(TreeNode root, int val) {
        if(root==null) return null;
        
        if(root.val==val) return root;
        
        return searchBST((val>root.val)? root.right : root.left, val);
        
    }